bundle/src/main/java/com/adobe/acs/commons/util/impl/DiscoveryServiceHelper.java

```diff
@@ -64,9 +59,6 @@ public void handleTopologyEvent(TopologyEvent event) {
     @Activate
     public void activate(BundleContext bundleContext) {
         this.bundleContext = bundleContext;
-        if (discoveryService.getTopology().getLocalInstance().isLeader()) {
```

Comment:
How often are relevant topology events send out? How much does this change potentially defer the `ClusterLeader` service from becoming active on a leader...

Reply:
They are sent as soon as a topology change is detected and has settled. 
There is no difference between getting the event vs asking the DiscoveryService - the underlying state is the same.

The problem with this activate method in addition btw is that the topology might not be defined yet - in which case TopologyView.isCurrent() would return false - ie the leader flag would be based on stale/not current information. If you rely on TopologyEventHandler this would not happen.

Reply:
I am not following here: In case `TopologyView.isCurrent()` would return false no `ClusterLeader` service is registered so I don't think there is a problem here.

> They are sent as soon as a topology change is detected and has settled.
The problem is rather that this service might start after a stable topology has been established already. In that case the `ClusterLeader` service would only be registered once a new topology change event is received.

The whole idea of registering the service eagerly in activate() is that otherwise it might take too long until this service figured out whether it runs on the leader or not.

Reply:
I was mentioning `TopologyView.isCurrent()` not `InstanceDescription.isLeader()`

Reply:
> The problem is rather that this service might start after a stable topology has been established already. In that case the ClusterLeader service would only be registered once a new topology change event is received.

Just to make sure we're talking about the same thing: there's both TOPOLOGY_CHANGED - but there's also TOPOLOGY_INIT. TOPOLOGY_INIT is for exactly this purpose: it sends the very first state of the topology (once it is defined) to the listener right/soon-after the listener was registered. Ie the listener doesn't have to wait for a topology change.

> The whole idea of registering the service eagerly in activate() is that otherwise it might take too long until this service figured out whether it runs on the leader or not.

Ideally we're talking a few millis

Reply:
Ok, but that could be easily fixed by adding another condition. Your change affects functionality in a bad way as it defers registering `ClusterLeader`, I am just not sure by how much...

Reply:
> TOPOLOGY_INIT is for exactly this purpose: it sends the very first state of the topology (once it is defined) to the listener right/soon-after the listener was registered. Ie the listener doesn't have to wait for a topology change.

Thanks, indeed that was the missing part. Then the change looks good to me.

Reply:
Ok great. I was gonna add that the difference is miniscule. In case of a standalone instance it would be merely one
[thread-switch](https://github.com/apache/sling-org-apache-sling-discovery-commons/blob/a484a7c9e70847831a41ac3b1e6168c786fd6d0c/src/main/java/org/apache/sling/discovery/commons/providers/base/ViewStateManagerImpl.java#L203) (as TopologyEvents are always sent asynchronously to avoid blocking) - and in case of a cluster it is likely that at activation time the topology hasn't even settled, so no time would be lost at all.